File: MPB/DBR_TMM_vs_PWE.m
close all;
clear all;

% load the TMM data
load('TMM_data.mat');
% load the MPB data
MPB_data = MPB_load_data('example_DBR_final.dat');

% get the gap infos
gap_infos = MPB_getGaps(MPB_data);

% set the Y axis range
y_limits = [0,0.7];

% create the figure
figure;

% plot the TMM data
subplot(1,2,1); hold on;
title('TMM');
plot(R, a./wvl_list, 'DisplayName', 'no defect');
ylabel('a/\lambda (no unit)');
xlabel('Reflectance R (no unit)');
hline(gap_infos.fullgaps.bottom, 'r--');
hline(gap_infos.fullgaps.top, 'b--');
ylim(y_limits);

% plot the MPB data
subplot(1,2,2); hold on;
title('PWE');
plot(MPB_data.k1, MPB_data.fn(:,1), 'r-');
plot(MPB_data.k1, MPB_data.fn(:,2), 'g-');
plot(MPB_data.k1, MPB_data.fn(:,3), 'b-');
xlabel('kx/(2\pi/a)');
ylabel('a/\lambda');
ylim(y_limits);
hline(gap_infos.fullgaps.bottom, 'r--');
hline(gap_infos.fullgaps.top, 'b--');

% save the figure
saveas(gcf, 'DBR_TMM_vs_PWE.svg');

File: MPB/DBR_TMM_vs_PWE.py
from tmm.tmm_core import coh_tmm
import numpy as np
import matplotlib.pyplot as plt

# DBR parameters
n1 = 1
d1 = 0.16 # um
n2 = 3.2
d2 = 0.16 # um
Nperiod = 15
a = d1 + d2

# list of layer thicknesses in nm
d_list = [np.inf] + Nperiod*[d1, d2] + [np.inf]
# list of refractive indices
n_list = [1] + Nperiod*[n1, n2] + [1]

# normalized frequency range
fn = np.linspace(0, 0.7, 1000)
# wavelength range
wvl_list = a/fn

# run TMM
R = []
for wvl in wvl_list:
    ret = coh_tmm('s', n_list, d_list, 0, wvl)
    R.append(ret['R'])

# plot R as a function of a/lambda
plt.plot(fn, R, label='no defect')
plt.xlabel(r'$a/\lambda$ (no unit)')
plt.ylabel('Reflectance $R$ (no unit)')
plt.title('Reflectance at $0\degree$ incidence')

# Save the data to .mat file that is easy to load in Matlab.
from scipy.io import savemat
mdic = {"wvl_list": wvl_list,
        "R": R,
         'n1': n1,
         'd1': d1,
         'n2': n2,
         'd2': d2,
         'Nperiod': Nperiod,
         'a': a}
savemat("TMM_data.mat", mdic)
